runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/processing/StreamingWorkScheduler.java

```diff
@@ -0,0 +1,413 @@
+import org.apache.beam.runners.dataflow.worker.windmill.Windmill;
+import org.apache.beam.runners.dataflow.worker.windmill.client.commits.Commit;
+import org.apache.beam.runners.dataflow.worker.windmill.state.WindmillStateReader;
+import org.apache.beam.runners.dataflow.worker.windmill.work.processing.failures.FailureTracker;
+import org.apache.beam.runners.dataflow.worker.windmill.work.processing.failures.WorkFailureProcessor;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.vendor.grpc.v1p60p1.com.google.protobuf.ByteString;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Schedules execution of user code to process a {@link
+ * org.apache.beam.runners.dataflow.worker.windmill.Windmill.WorkItem} then commits the work item
+ * back to streaming execution backend.
+ */
+@Internal
+@ThreadSafe
+public final class StreamingWorkScheduler {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingWorkScheduler.class);
+
+  private final DataflowWorkerHarnessOptions options;
+  private final Supplier<Instant> clock;
+  private final ExecutionStateFactory executionStateFactory;
+  private final SideInputStateFetcher sideInputStateFetcher;
+  private final FailureTracker failureTracker;
+  private final WorkFailureProcessor workFailureProcessor;
+  private final StreamingCommitFinalizer commitFinalizer;
+  private final StreamingCounters streamingCounters;
+  private final HotKeyLogger hotKeyLogger;
+  private final ConcurrentMap<String, StageInfo> stageInfoMap;
+  private final DataflowExecutionStateSampler sampler;
+  private final AtomicInteger maxWorkItemCommitBytes;
+
+  public StreamingWorkScheduler(
+      DataflowWorkerHarnessOptions options,
+      Supplier<Instant> clock,
+      ExecutionStateFactory executionStateFactory,
+      SideInputStateFetcher sideInputStateFetcher,
+      FailureTracker failureTracker,
+      WorkFailureProcessor workFailureProcessor,
+      StreamingCommitFinalizer commitFinalizer,
+      StreamingCounters streamingCounters,
+      HotKeyLogger hotKeyLogger,
+      ConcurrentMap<String, StageInfo> stageInfoMap,
+      DataflowExecutionStateSampler sampler,
+      AtomicInteger maxWorkItemCommitBytes) {
+    this.options = options;
+    this.clock = clock;
+    this.executionStateFactory = executionStateFactory;
+    this.sideInputStateFetcher = sideInputStateFetcher;
+    this.failureTracker = failureTracker;
+    this.workFailureProcessor = workFailureProcessor;
+    this.commitFinalizer = commitFinalizer;
+    this.streamingCounters = streamingCounters;
+    this.hotKeyLogger = hotKeyLogger;
+    this.stageInfoMap = stageInfoMap;
+    this.sampler = sampler;
+    this.maxWorkItemCommitBytes = maxWorkItemCommitBytes;
+  }
+
+  private static ShardedKey createShardedKey(Work work) {
```

Comment:
should Work contain a ShardedKey instead?

Reply:
done